### ISSUE TITLE
Checkout API can create a user account when order fails due to coupon restriction (or other failures)

### ISSUE THREAD

**Issue by haszari** [woocommerce/woocommerce-gutenberg-products-block#3363]
## Describe the bug
~~This affects stores with no payment methods set up, with checkout block configured to allow sign-up on checkout. If the shopper checks `Create an account` and submits the order, the account is created, even though the order isn't processed because of no valid payment method.~~

There may be similar issues, e.g. if a payment fails.

**Update** Refocusing this bug on a specific case for now. When a restricted coupon is used, an account is created and associated with the draft order. 

## To reproduce
See [comment below for test setup & steps](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/3363#issuecomment-728387054).

Previous "No payment methods" steps:

1. Add a checkout block to a published page with Allow shoppers to sign up for a user account during checkout option enabled.
2. In `WooCommerce > Settings > Payments` disable all payment methods.
4. On front end, logged out (or incognito) add something to cart and proceed to checkout.
4. Complete checkout form and select Create an account.
4. Submit.

## Expected behavior
~~No account created unless order succeeds. The order processing and any related logic/entities should only happen if the order succeeds (or preflight checks confirm that the order will succeed.~~

- Account should only be created when associated with a successful order. No account should be created for `checkout-draft` order status.
- Store API should behave predictably and consistently (TBD!) when creating accounts and progressing order state. There may be some differences with legacy / shortcode checkout processing.

**Comment by senadir** [comment]
I did some digging around for the next person who wants to solve this:

### How Core does it:

- Core validates an order from posted_data before processing anything in `WC_Checkout::validate_checkout`.
- It creates a user after doing that.

### How we do it:

- We create an account before doing any order validation.
- Our validation is done in `OrderController::validate_order_before_payment` and does validation based on the passed order.
- That validation only includes coupon validation and is tightly coupled to it.
- We later (when processing payment) attempt to validate the payment gateway.

We're going to run into an issue or two:
- We're either going to hit `$order->save()` twice, which I believe is a heavy operation cc @mikejolley?
- We need to move `$order->save()` from outside `update_order_from_request` and call it back a bit further down.

**Comment by mikejolley** [comment]
Why does account creation need to be coupled to order success? If future payment requests do not fail or attempt to create another account..

**Comment by senadir** [comment]
That's a good question @mikejolley 
This is mostly the status quo when it comes to Core, should we follow it or not is the question.
Problems I can think of are security and probably membership issues in which only paid users can have an account, changing this known behaviour in Woo here might be problematic and cause backward compatibility issues.
I'd prefer if we can capture this data without creating an account.

**Comment by haszari** [comment]
> Why does account creation need to be coupled to order success? If future payment requests do not fail or attempt to create another account..

Good question @mikejolley  - I hastily reported this bug when I spotted it, but perhaps we can iterate a bit more on the specifics of the problem (and what fixes might be needed). 

~~I just tested the shortcode checkout for comparison. It doesn't allow order submit when there are no payment methods. So that potentially masks this issue, and is possibly a bug with the checkout block.~~ Edit: not sure how I saw this, can't reproduce. #3411

Regarding the back-end handling of new accounts when the order fails, I tested that too. To test this I hacked BACS payment method in core so it always fails (return fail early, see code below). When this happens the order is created, the user account is created, and the order is in `Pending Payment` status.

```php
	public function process_payment( $order_id ) {
		return array(
			'result'   => 'fail',
			'redirect' => '',
		);
```

<img width="674" alt="Screen Shot 2020-11-17 at 9 34 55 AM" src="https://user-images.githubusercontent.com/4167300/99305093-300a3180-28b8-11eb-8928-6dabd74cb1bc.png">

So it looks like maybe this isn't a bug after all - perhaps it's consistent with legacy checkout, in line with your question 😄. There might be other quirks with different edge cases, or we might want stricter behaviour for Store API, but perhaps we don't need to do anything.

**Comment by nerrad** [comment]
> It doesn't allow order submit when there are no payment methods. So that potentially masks this issue, and is possibly a bug with the checkout block. I'll log that separately: #3411

Relying on a ui restriction as a fix for this is problematic imo. #3411 would still be good to address from a UX perspective, but technically it's probably easier to trigger the issue reported here via the StoreAPI directly than whatever exists in Woo Core right now. So I think this is still something that needs considered. Let's recap:

- Currently the shortcode flow doesn't have the concept of draft orders and order _creation_ is conditional on being triggered from the checkout submit. The only time an account is not created is when the checkout posted data does not validate (not contingent on payment). Thus account creation is coupled to _order creation_ (not payment).
- The block checkout flow always has a draft order, so orders are _always_ created in the flow. Thus having the coupling of account creation to order creation could be problematic here because we don't want an account created for _draft orders only_ (this could result in "orphaned" user accounts in the store if the order never changes from the `checkout-draft` state and the order gets purged).

Since there is a fundamental difference in _order creation_ between the two flows, I think this issue still needs addressed.

**Comment by haszari** [comment]
> The block checkout flow always has a draft order, so orders are always created in the flow. Thus having the coupling of account creation to order creation could be problematic here because we don't want an account created for draft orders only

I agree @nerrad - that's a key difference here, thanks for clarifying. I'll work on fixing the API so that this doesn't happen. i.e. accounts will only be created after the order is validated and has successful payment.

**Comment by senadir** [comment]
Note that this might be a breaking change for `__experimental_woocommerce_blocks_checkout_order_processed` because Subscriptions expects account to be already created at that point, and also needs the cart to be full. Processing payments empties the cart. So we need to trigger that action after account creation and validation but just before processing payment.

Apart from running side code, that action can also be useful for adding extra validation to the cart.

**Comment by haszari** [comment]
> Note that this might be a breaking change for __experimental_woocommerce_blocks_checkout_order_processed because Subscriptions expects account to be already created at that point, and also needs the cart to be full. Processing payments empties the cart. So we need to trigger that action after account creation and validation but just before processing payment.

Good point @senadir. I'm going to do more testing to confirm what scenarios are broken (if any) before attempting a fix. 

The order of processing in `Checkout::get_route_post_response()` is super critical as you've suggested, so I wouldn't want to reorder it without good reason (and extensive testing).

**Comment by haszari** [comment]
I tested a few different scenarios and found a few differences between shortcode vs block, and some bugs to fix. 

Here's a summary:

Condition | Order? | Account | Bug?
--- | --- | --- | --- 
**Shortcode / Legacy** | | |
No payment methods | no order | no account | 
Payment method hacked to fail (COD) | `pending` | account |
Coupon restriction order fail | no order | no account |
No stock order fail | no order | no account |
**Block / StoreAPI** | | | |
No payment methods | `pending` | account |
Payment method hacked to fail (COD) | `pending` | account | 
Coupon restriction order fail | `checkout-draft` | account | 🐛 #3363 (this issue)
No stock order fail | `pending` | account  | 🐛 #3412

### Outcomes from testing
1. We have a difference of behaviour for `No payment methods` case, though this may be ok; TBD, maybe we should aim to match existing behaviour, or be more strict (user account only created if order succeeds).
2. Coupon failure produces a draft order associated with a new customer account. This is a bug, and I'm going to refocus this issue on this specific problem.
3. We have an unrelated bug that orders succeed when item is out of stock. Logging this separately.

Initially I'm going to look at the coupon fail draft order issue. 

#### Some details about how I tested
Configuration:

- `Allow create account` = YES in store settings and block.
- `Allow guest checkout` = YES, `Allow checkout log in` = NO
- COD payment method only (unless testing "No payment methods" condition).
- COD payment fail was implemented by returning early in COD `process_payment` with `result=> fail`. This is intended to simulate any payment failure server-side.
- "Coupon restriction" - Coupon only allows specific email host, orders submitted with invalid host for coupon.
- "No stock" - Before clicking place order, in back end I update the product so it is out of stock.

Test steps:

- Ensure logged out.
- Add product to cart.
- View checkout (block or shortcode as appropriate to test – I have two pages set up).
- Enter a unique email (i.e. doesn't already have an account) and complete checkout form.
- Click place order.

**Comment by haszari** [comment]
I'm starting to think it's cleaner to avoid making an account if the order fails for other reasons (e.g. payment method related). In this case, the flow is arguably broken:

- Guest shopper selects `Create an account` and submits order.
- Account is created, user is logged in.
- Payment fails server side (e.g. no payment methods, or other reasons).
- Checkout block renders error to user.

At this point the checkout is showing the UI for a logged-out user (`Create an account` is visible) but the user is logged in.

When testing #3423 I saw this behaviour and perhaps it's worth a follow up - either make the checkout aware of login change from AJAX request, or only create an account if order succeeds through all stages including payment. 

<img width="944" alt="Screen Shot 2020-11-18 at 5 51 10 PM" src="https://user-images.githubusercontent.com/4167300/99485450-5b2f7680-29c7-11eb-88d4-2af7a0fd6dd8.png">

Note there is a store API nonce error in this scenario (unrelated bug).

**Comment by haszari** [comment]
Update: I've got a fix for the coupon issue in #3423. I'm interested in feedback on whether that fix closes this issue, and/or whether we want to look at changing the behaviour for when an order fails due to payment. cc @nerrad @mikejolley (Store API)

**Comment by mikejolley** [comment]
> Note there is a store API nonce error in this scenario (unrelated bug).

Just to note this might not be totally unrelated. Logged in users would get a different nonce, so that might be the reason this is turning invalid. We need to ensure a new updated nonce is returned if an account is created via the API.

**Comment by haszari** [comment]
Aha! Good point @mikejolley. Sounds like we could address both ends of this:

- Avoid creating account server-side unless order is definitely going to succeed.
- Improve client-side AJAX handling of when user is logged in/nonce changes - we should be robust to this really (even if we prevent the issue server side).

**Comment by haszari** [comment]
> Improve client-side AJAX handling of when user is logged in/nonce changes - we should be robust to this really (even if we prevent the issue server side).

Working on this in a separate PR #3429
